packages/amd-loader/amd-loader.js

```diff
@@ -0,0 +1,196 @@
+'use strict';
+
+(function() {
+
+function Registry() {
+  this.modules = {};
+  this.topLevelIdx = 0;
+  this.previousTopLevelUrl = undefined;
+  this.pendingDefine = undefined;
+}
+
+Registry.prototype.get = function(url) {
+  var module = this.modules[url];
+  if (module === undefined) {
+    module = this.modules[url] = new Module(url);
+  }
+  return module;
+};
+
+var registry = new Registry();
+
+function getBasePath(url) {
+  // TODO(aomarks) Maybe a better way to do this.
+  var lastSlash = url.lastIndexOf('/');
+  if (lastSlash === url.indexOf('://') + 2) {
+    // http://example.com -> http://example.com/
+    return url + '/';
+  }
+  // http://example.com/foo/bar.html -> http://example.com/foo/
+  return url.substring(0, lastSlash + 1);
+}
+
+function Module(url) {
+  this.url = url;
+  this.basePath = getBasePath(url);
+  this.exports = {};
+  this.loadStartedOrDone = false;
+  // All of this module's dependencies are resolved and its factory has run.
+  this.resolved = false;
+  // Callbacks from dependents waiting to hear when this module has resolved.
+  this.notify = [];
+}
+
+var resolverAnchor = document.createElement('a');
+
+Module.prototype.init = function(deps, factory) {
+  var factoryArgs = [];
+  var numUnresolvedDeps = 0;
+
+  function onDepResolved() {
+    numUnresolvedDeps--;
+    maybeResolve();
+  }
+
+  var thisResolve = this.resolve.bind(this);
+  function maybeResolve() {
+    if (numUnresolvedDeps === 0) {
+      factory.apply(null, factoryArgs);
+      thisResolve();
+    }
+  }
+
+  for (var i = 0; i < (deps || []).length; i++) {
+    var depSpec = deps[i];
+
+    if (depSpec === 'exports') {
+      factoryArgs[i] = this.exports;
+
+    } else if (depSpec === 'require') {
+      factoryArgs[i] = this.makeDynamicImporter();
+
+    } else if (depSpec === 'meta') {
+      // TODO(aomarks) This is an idea for replacing the import.meta transform.
+      factoryArgs[i] = {url: this.url};
+
+    } else {
+      var depModule = registry.get(this.resolveUrl(depSpec));
+      factoryArgs[i] = depModule.exports;
+
+      if (depModule.resolved === false) {
+        numUnresolvedDeps++;
+        depModule.notify.push(onDepResolved);
+        depModule.loadOnce();
+      }
+    }
+  }
+
+  maybeResolve();
+};
+
+Module.prototype.resolve = function() {
+  this.resolved = true;
+  for (var i = 0; i < this.notify.length; i++) {
+    this.notify[i]();
+  }
+};
+
+Module.resolverAnchor = document.createElement('a');
+
+Module.prototype.resolveUrl = function(path) {
+  if (path.indexOf('://') !== -1) {
+    // Already a fully qualified URL.
+    // TODO(aomarks) Is this a good enough check?
```

Comment:
If you have `URL`, then I think you can try-catch `new URL(path)`, which throws if it's not given a full URL, IIRC.